### PR TITLE
LRUcache refactor

### PR DESCRIPTION
**Pull request by elainexmas**
DRY up existing cache classes using inheritance.

Still need to add time-limited cache type.  This PR just addresses existing cache types.

### CODE REVIEW

File: R/lru.R [hunk 1]
```diff
@@ -30,21 +20,26 @@ LRUcache_class <- R6::R6Class("LRUcache",
...
     max_num = 0,
     data    = new.env(),
-    use_bytes = FALSE,
     save    = function(name, value) {
       # Check if this value alone exceeds the cache size
-      size <- if(private$use_bytes) as.integer(as.character(pryr::object_size(value))) else 1
+      size <- private$get_new_item_size(value)
       if (size > private$max_num){
-        warning(sprintf("'%s' is too large (%dB) to fit in the cache (%dB) and will not be cached. Consider creating a larger cache.", name, size, private$max_num), call. = FALSE)
+        warning(sprintf("'%s' is too large (%d) to fit in the cache (%d) and will not be cached. Consider creating a larger cache.", name, size, private$max_num), call. = FALSE)
```
Comment (peterhurford) [inline R/lru.R]: I liked specifying that it was bytes.

Comment (elainexmas) [inline R/lru.R]: Ah yes, I did like that too.  I'll bring it back, even if it means an if-else to turn on the units if it is a memory size (`'character'`) cache and off if it's just a number-of-items (`'numeric'`) cache.

**Comment by elainexmas** [comment]
@peterhurford @kirillseva Any other comments?

File: R/lru.R [hunk 2]
```diff
@@ -30,21 +20,29 @@ LRUcache_class <- R6::R6Class("LRUcache",
...
-    use_bytes = FALSE,
+    units   = '',
     save    = function(name, value) {
       # Check if this value alone exceeds the cache size
-      size <- if(private$use_bytes) as.integer(as.character(pryr::object_size(value))) else 1
+      size <- private$get_new_item_size(value)
       if (size > private$max_num){
-        warning(sprintf("In package cacher: '%s' is too large (%dB) to fit in the LRU cache (%dB) and will not be cached. Consider creating a larger cache.", name, size, private$max_num), call. = FALSE)
+        warning(sprintf("'%s' is too large (%d%s) to fit in the cache (%d%s) and will
+          not be cached. Consider creating a larger cache.",
```
Comment (peterhurford) [inline R/lru.R:40]: Can you revert back to the old message?

Comment (peterhurford) [inline R/lru.R:40]: I like the units part though.

**Comment by peterhurford** [comment]
Looks good to me but I don't really know that much about R6 :+1: 

Any changes or additions that need to go into test coverage?

**Comment by kirillseva** [comment]
Hi! Thanks for the PR and sorry for being so unresponsive. I'll take a look as soon as I get home!

File: R/lru.R [hunk 3]
```diff
@@ -1,16 +1,6 @@
-LRUcache_class <- R6::R6Class("LRUcache",
+LRUcache_ <- R6::R6Class("LRUcache_",
   public = list(
     initialize = function(size) {
```
Comment (kirillseva) [inline R/lru.R:3]: hmm, I think you need to set `private$max_num <- size` ?

File: R/lru.R [hunk 4]
```diff
@@ -30,21 +20,29 @@ LRUcache_class <- R6::R6Class("LRUcache",
       stopifnot(is.character(name) && length(name) == 1)
       stopifnot(name %in% ls(private$data))
       private$peek(name)
+    },
+    print = function() {
```
Comment (kirillseva) [inline R/lru.R:24]: this is awesome

File: R/lru.R (hunk 3 above)
Comment (kirillseva) [inline R/lru.R:3]: I see, it's defined in subclasses

File: R/lru.R [hunk 5]
```diff
@@ -61,6 +59,46 @@ LRUcache_class <- R6::R6Class("LRUcache",
...
+LRUcache_.numeric <- R6::R6Class("LRUcache_.numeric", inherit = LRUcache_,
+  public = list(
+    initialize = function(size) {
+      stopifnot(all.equal(size, as.integer(size)) && length(size) == 1 && size > 0)
+      private$max_num = size
+    }
+  ),
+  private = list(
+    get_new_item_size = function(item) 1,
+    check_for_eviction = function(name, size){
```
Comment (kirillseva) [inline R/lru.R]: could you add a space between `)` and `{` please? :beetle:

Comment (kirillseva) [inline R/lru.R]: could you add space between `(` and `{` ?

File: R/lru.R [hunk 6]
```diff
@@ -61,6 +59,46 @@ LRUcache_class <- R6::R6Class("LRUcache",
       oldest <- names(which.min(times))
       rm(list = oldest, envir = private$data)
     },
+    format_cache = function(){
```
Comment (kirillseva) [inline R/lru.R]: please add a space before `{` and reformat the body of the function. The line is too wide and it's confusing

File: R/lru.R [hunk 7]
```diff
@@ -61,6 +59,46 @@ LRUcache_class <- R6::R6Class("LRUcache",
...
+      stopifnot(all.equal(size, as.integer(size)) && length(size) == 1 && size > 0)
+      private$max_num = size
+    }
+  ),
+  private = list(
+    get_new_item_size = function(item) 1,
+    check_for_eviction = function(name, size){
+      private$get_current_size() + size > private$max_num && !(name %in% ls(private$data))
+    },
+    get_current_size = function(){
```
Comment (kirillseva) [inline R/lru.R]: same thing with spacing

File: R/lru.R (hunk 5 above)
Comment (kirillseva) [inline R/lru.R]: Looks like `check_for_eviction` method can be moved to the parent class?

**Comment by kirillseva** [comment]
This is awesome! Thank you so much for this PR @elainexmas!

Just a couple small things I'd like to get addressed before merging

**Comment by robertzk** [comment]
I like it.

**Comment by kirillseva** [comment]
So do I!

**Pull request merged**